src/replication.c

```diff
@@ -98,6 +98,67 @@ unsigned long replicationLogicalReplicaCount(void) {
     return count;
 }
 
+int replicaFromIOThreadHasPendingAck(client *c) {
+    serverAssert(c->tid != IOTHREAD_MAIN_THREAD_ID);
+
+    int pending_read;
+    atomicGetWithSync(c->pending_read, pending_read);
+    return pending_read;
+}
+
+/* Send replicas to their respective IO threads if it has pending reads or
+ * writes. Otherwise it remains in main thread so it can check for new data in
+ * the replication buffer ASAP. */
+void putReplicasInPendingClientsToIOThreads(void) {
+    if (server.io_threads_num <= 1) return;
+
+    serverAssert(pthread_equal(pthread_self(), server.main_thread_id));
+
+    listIter li;
+    listNode *ln;
+    listRewind(server.slaves,&li);
+    while((ln = listNext(&li))) {
+        client *slave = ln->value;
+
+        if (slave->tid == IOTHREAD_MAIN_THREAD_ID ||
+            slave->running_tid != IOTHREAD_MAIN_THREAD_ID)
+        {
+            continue;
+        }
+ 
+        if (slave->flags & CLIENT_PENDING_WRITE ||
```

Comment:
maybe for slave, we can check if it has pending replication stream to send, instead of just relying on `CLIENT_PENDING_WRITE`, so we don't need a tricky  in `_prepareClientToWrite `